### PR TITLE
Core: Extract REST metrics reporter into its own class

### PR DESCRIPTION
**Pull request by nastra**
depends on https://github.com/apache/iceberg/pull/7337

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      } else {
+        LOG.warn("Failed to derive namespace/table name from metrics report {}", report);
+      }
+
+      if (null != tableIdentifier) {
+        client.post(
+            paths.metrics(tableIdentifier),
+            ReportMetricsRequest.of(report),
+            null,
+            ImmutableMap.of(),
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:70]: it's worth mentioning that this doesn't use any authorization headers to send a metrics report to the REST endpoint

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    List<String> split = DOT.splitToList(tableNameWithCatalog);
+    if (split.size() > 2) {
+      return TableIdentifier.parse(tableNameWithCatalog.replace(split.get(0) + ".", ""));
+    }
+
+    return TableIdentifier.parse(tableNameWithCatalog);
+  }
+
+  @Override
+  public void close() {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:89]: we might want to consider making `CompositeMetricsReporter` implement `Closeable` as well and propagate closing through it

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.iceberg.metrics.MetricsReport;
+import org.apache.iceberg.metrics.MetricsReporter;
+import org.apache.iceberg.metrics.ScanReport;
+import org.apache.iceberg.relocated.com.google.common.base.Splitter;
+import org.apache.iceberg.rest.requests.ReportMetricsRequest;
+import org.apache.iceberg.util.SerializableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RESTMetricsReporter implements MetricsReporter {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:34]: making `RESTMetricsReporter` fully `Serializable` will be handled in a separate PR as it currently also requires https://github.com/apache/iceberg/pull/7144

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.iceberg.metrics.CommitReport;
+import org.apache.iceberg.metrics.MetricsReport;
+import org.apache.iceberg.metrics.MetricsReporter;
+import org.apache.iceberg.metrics.ScanReport;
+import org.apache.iceberg.relocated.com.google.common.base.Splitter;
+import org.apache.iceberg.rest.requests.ReportMetricsRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class RESTMetricsReporter implements MetricsReporter {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:33]: Serializable? Or is this just to produce a separate one as a minimal refactor?

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 5]
```diff
@@ -202,8 +200,12 @@ public void initialize(String name, Map<String, String> unresolved) {
 
     this.reporter = CatalogUtil.loadMetricsReporter(mergedProps);
 
-    this.reportingViaRestEnabled =
-        PropertyUtil.propertyAsBoolean(mergedProps, REST_METRICS_REPORTING_ENABLED, true);
+    if (PropertyUtil.propertyAsBoolean(mergedProps, REST_METRICS_REPORTING_ENABLED, true)) {
+      this.reporter =
+          MetricsReporters.combine(
+              this.reporter, new RESTMetricsReporter(client, paths, catalogAuth.headers()));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:206]: Style: I think it would be better to avoid all of the complex formatting by moving `new RESTMetricsReporter(...)` to its own statement.

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(RESTMetricsReporter.class);
+  private static final Splitter DOT = Splitter.on('.');
+
+  private final RESTClient client;
+  private final ResourcePaths paths;
+  private final Map<String, String> headers;
+
+  RESTMetricsReporter(RESTClient client, ResourcePaths paths, Map<String, String> headers) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:41]: Won't the headers change when the auth session is refreshed?

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 7]
```diff
@@ -127,7 +127,7 @@ public <T extends RESTResponse> T execute(
             // this doesn't use a Mockito spy because this is used for catalog tests, which have
             // different method calls
             if (!"v1/oauth/tokens".equals(path)) {
-              if ("v1/config".equals(path)) {
+              if ("v1/config".equals(path) || path.endsWith("/metrics")) {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:130]: Why was this needed?

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            ImmutableMap.of(),
+            ErrorHandlers.defaultErrorHandler());
+      }
+    } catch (Exception e) {
+      LOG.warn("Failed to report metrics to REST endpoint for table {}", tableIdentifier, e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private TableIdentifier identifierWithoutCatalog(String tableNameWithCatalog) {
```
[the remark this answers is not in the document]
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:78]: It seems like a bad idea to parse this. Can we add the identifier to the ScanReport instead? Then we could use the identifier that was loaded. Maybe we should add it to `BaseTable` as well, like the Python implementation does.

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java (hunk 6 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:41]: ah yes, this should be a `Supplier<Map<String, String>>`

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java (hunk 4 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:33]: for easier reviewing, the goal of this PR was to move REST metrics reporting into its own class before making it fully serializable (as that involves slightly more changes). I have a [WIP PR](https://github.com/apache/iceberg/pull/7370) for this that also includes the required changes from #7144

File: core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTMetricsReporter.java:78]: I completely agree here. It would be great if the actual table would carry a `TableIdentifier`, then we could just use that. I'll open a separate issue to improve that.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 9]
```diff
@@ -347,35 +346,26 @@ public Table loadTable(SessionContext context, TableIdentifier identifier) {
...
-        client.post(
-            paths.metrics(tableIdentifier),
-            ReportMetricsRequest.of(report),
-            null,
-            headers,
-            ErrorHandlers.defaultErrorHandler());
-      }
-    } catch (Exception e) {
-      LOG.warn("Failed to report metrics to REST endpoint for table {}", tableIdentifier, e);
+  private MetricsReporter metricsReporter(
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:361]: rather than wrapping the combined reporter in a lambda, we just return the combined reporter here.

**Review by rdblue — APPROVED** [review]
